src/Microsoft.AspNetCore.SignalR.Core/SerializedHubMessage.cs

```diff
@@ -80,21 +124,32 @@ private void SetCache(string protocolName, ReadOnlyMemory<byte> serialized)
             }
         }
 
-        private bool TryGetCached(string protocolName, out ReadOnlyMemory<byte> result)
+        private bool TryGetCachedFast(string protocolName, out ReadOnlyMemory<byte> result)
         {
-            if (string.Equals(_cachedItem1.ProtocolName, protocolName, StringComparison.Ordinal))
+            if (_count > 0 && string.Equals(_cachedItem1.ProtocolName, protocolName, StringComparison.Ordinal))
```

Comment:
Add a comment that the count field is used to avoid problems with the struct sheering

Reply:
👍 

Though that made me thing of another thing: Do we need to make the fields `volatile` to ensure read/write barriers are used? Could out-of-order execution hit us here? I don't remember the specifics of that :(. @davidfowl ?

Reply:
Sounds like @stephentoub confirmed my fears, that `_count` could be observed to be updated from the other thread before `_cachedItem` is observed to be updated. We're going to remove the "fast" path for now (we had locking here earlier, before my refactor that broke it). If we can get a correct lock-free path for reading the cache we can optimize later.